File: nc-150/trees/depth_of_binary_tree.cpp
/*
 * Author: Valerie Doan
 * Date: October 27, 2024
 * Problem: Maximum Depth of Binary Tree
 * Level: Easy
 * Neetcode Course: Neetcode 150
 * Neetcode Topic: Trees
 */

/*
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

void helper(TreeNode *root, int &res, int height)
{
    if (!root)
    {
        return;
    }
    res = max(res, ++height);
    helper(root->left, res, height);
    helper(root->right, res, height);
}

int maxDepth(TreeNode *root)
{
    int res = 0;
    helper(root, res, 0);
    return res;
}

File: python/nc-150/trees/depth_of_binary_tree.py
"""
Author: Valerie Doan
Date: November 7, 2024
Problem: Maximum Depth of Binary Tree
Level: Easy
Neetcode Course: Neetcode 150
Neetcode Topic: Trees
"""


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        left = self.maxDepth(root.left) + 1
        right = self.maxDepth(root.right) + 1
        return max(left, right)
